From: Sergio Martin
Subject: [PATCH] chimera: players 2 to 4 have keys by default

Upstream gives players 2 to 4 no controls by default (clearbuttons), and
its defines for them name the first joysticks. The sandbox has no
joysticks, so they get keys, as player 1 has: the core presses those keys
for its player 2 to 4 controls.
---
diff --git a/engine/openbor.c b/engine/openbor.c
index ba010e9c..e6b63d5d 100644
--- a/engine/openbor.c
+++ b/engine/openbor.c
@@ -2838,17 +2838,17 @@ void clearbuttons(int player)
     }
     else if (player == 1)
     {
-        savedata.keys[1][SDID_MOVEUP]    = CONTROL_NONE; //Kratus (20-04-21) Used to clear all keys
-        savedata.keys[1][SDID_MOVEDOWN]  = CONTROL_NONE;
-        savedata.keys[1][SDID_MOVELEFT]  = CONTROL_NONE;
-        savedata.keys[1][SDID_MOVERIGHT] = CONTROL_NONE;
-        savedata.keys[1][SDID_ATTACK]    = CONTROL_NONE;
-        savedata.keys[1][SDID_ATTACK2]   = CONTROL_NONE;
-        savedata.keys[1][SDID_ATTACK3]   = CONTROL_NONE;
-        savedata.keys[1][SDID_ATTACK4]   = CONTROL_NONE;
-        savedata.keys[1][SDID_JUMP]      = CONTROL_NONE;
-        savedata.keys[1][SDID_SPECIAL]   = CONTROL_NONE;
-        savedata.keys[1][SDID_START]     = CONTROL_NONE;
+        savedata.keys[1][SDID_MOVEUP]    = CONTROL_DEFAULT2_UP; // chimera: keys, as player 1 has
+        savedata.keys[1][SDID_MOVEDOWN]  = CONTROL_DEFAULT2_DOWN; // chimera: keys, as player 1 has
+        savedata.keys[1][SDID_MOVELEFT]  = CONTROL_DEFAULT2_LEFT; // chimera: keys, as player 1 has
+        savedata.keys[1][SDID_MOVERIGHT] = CONTROL_DEFAULT2_RIGHT; // chimera: keys, as player 1 has
+        savedata.keys[1][SDID_ATTACK]    = CONTROL_DEFAULT2_FIRE1; // chimera: keys, as player 1 has
+        savedata.keys[1][SDID_ATTACK2]   = CONTROL_DEFAULT2_FIRE2; // chimera: keys, as player 1 has
+        savedata.keys[1][SDID_ATTACK3]   = CONTROL_DEFAULT2_FIRE3; // chimera: keys, as player 1 has
+        savedata.keys[1][SDID_ATTACK4]   = CONTROL_DEFAULT2_FIRE4; // chimera: keys, as player 1 has
+        savedata.keys[1][SDID_JUMP]      = CONTROL_DEFAULT2_FIRE5; // chimera: keys, as player 1 has
+        savedata.keys[1][SDID_SPECIAL]   = CONTROL_DEFAULT2_FIRE6; // chimera: keys, as player 1 has
+        savedata.keys[1][SDID_START]     = CONTROL_DEFAULT2_START; // chimera: keys, as player 1 has
         savedata.keys[1][SDID_SCREENSHOT] = CONTROL_NONE;
         #ifdef SDL
             //savedata.keys[1][SDID_ESC]       = CONTROL_DEFAULT2_ESC;
@@ -2856,17 +2856,17 @@ void clearbuttons(int player)
     }
     else if (player == 2)
     {
-        savedata.keys[2][SDID_MOVEUP]    = CONTROL_NONE; //Kratus (20-04-21) Used to clear all keys
-        savedata.keys[2][SDID_MOVEDOWN]  = CONTROL_NONE;
-        savedata.keys[2][SDID_MOVELEFT]  = CONTROL_NONE;
-        savedata.keys[2][SDID_MOVERIGHT] = CONTROL_NONE;
-        savedata.keys[2][SDID_ATTACK]    = CONTROL_NONE;
-        savedata.keys[2][SDID_ATTACK2]   = CONTROL_NONE;
-        savedata.keys[2][SDID_ATTACK3]   = CONTROL_NONE;
-        savedata.keys[2][SDID_ATTACK4]   = CONTROL_NONE;
-        savedata.keys[2][SDID_JUMP]      = CONTROL_NONE;
-        savedata.keys[2][SDID_SPECIAL]   = CONTROL_NONE;
-        savedata.keys[2][SDID_START]     = CONTROL_NONE;
+        savedata.keys[2][SDID_MOVEUP]    = CONTROL_DEFAULT3_UP; // chimera: keys, as player 1 has
+        savedata.keys[2][SDID_MOVEDOWN]  = CONTROL_DEFAULT3_DOWN; // chimera: keys, as player 1 has
+        savedata.keys[2][SDID_MOVELEFT]  = CONTROL_DEFAULT3_LEFT; // chimera: keys, as player 1 has
+        savedata.keys[2][SDID_MOVERIGHT] = CONTROL_DEFAULT3_RIGHT; // chimera: keys, as player 1 has
+        savedata.keys[2][SDID_ATTACK]    = CONTROL_DEFAULT3_FIRE1; // chimera: keys, as player 1 has
+        savedata.keys[2][SDID_ATTACK2]   = CONTROL_DEFAULT3_FIRE2; // chimera: keys, as player 1 has
+        savedata.keys[2][SDID_ATTACK3]   = CONTROL_DEFAULT3_FIRE3; // chimera: keys, as player 1 has
+        savedata.keys[2][SDID_ATTACK4]   = CONTROL_DEFAULT3_FIRE4; // chimera: keys, as player 1 has
+        savedata.keys[2][SDID_JUMP]      = CONTROL_DEFAULT3_FIRE5; // chimera: keys, as player 1 has
+        savedata.keys[2][SDID_SPECIAL]   = CONTROL_DEFAULT3_FIRE6; // chimera: keys, as player 1 has
+        savedata.keys[2][SDID_START]     = CONTROL_DEFAULT3_START; // chimera: keys, as player 1 has
         savedata.keys[2][SDID_SCREENSHOT] = CONTROL_NONE;
         #ifdef SDL
             //savedata.keys[2][SDID_ESC]       = CONTROL_DEFAULT3_ESC;
@@ -2874,17 +2874,17 @@ void clearbuttons(int player)
     }
     else if (player == 3)
     {
-        savedata.keys[3][SDID_MOVEUP]    = CONTROL_NONE; //Kratus (20-04-21) Used to clear all keys
-        savedata.keys[3][SDID_MOVEDOWN]  = CONTROL_NONE;
-        savedata.keys[3][SDID_MOVELEFT]  = CONTROL_NONE;
-        savedata.keys[3][SDID_MOVERIGHT] = CONTROL_NONE;
-        savedata.keys[3][SDID_ATTACK]    = CONTROL_NONE;
-        savedata.keys[3][SDID_ATTACK2]   = CONTROL_NONE;
-        savedata.keys[3][SDID_ATTACK3]   = CONTROL_NONE;
-        savedata.keys[3][SDID_ATTACK4]   = CONTROL_NONE;
-        savedata.keys[3][SDID_JUMP]      = CONTROL_NONE;
-        savedata.keys[3][SDID_SPECIAL]   = CONTROL_NONE;
-        savedata.keys[3][SDID_START]     = CONTROL_NONE;
+        savedata.keys[3][SDID_MOVEUP]    = CONTROL_DEFAULT4_UP; // chimera: keys, as player 1 has
+        savedata.keys[3][SDID_MOVEDOWN]  = CONTROL_DEFAULT4_DOWN; // chimera: keys, as player 1 has
+        savedata.keys[3][SDID_MOVELEFT]  = CONTROL_DEFAULT4_LEFT; // chimera: keys, as player 1 has
+        savedata.keys[3][SDID_MOVERIGHT] = CONTROL_DEFAULT4_RIGHT; // chimera: keys, as player 1 has
+        savedata.keys[3][SDID_ATTACK]    = CONTROL_DEFAULT4_FIRE1; // chimera: keys, as player 1 has
+        savedata.keys[3][SDID_ATTACK2]   = CONTROL_DEFAULT4_FIRE2; // chimera: keys, as player 1 has
+        savedata.keys[3][SDID_ATTACK3]   = CONTROL_DEFAULT4_FIRE3; // chimera: keys, as player 1 has
+        savedata.keys[3][SDID_ATTACK4]   = CONTROL_DEFAULT4_FIRE4; // chimera: keys, as player 1 has
+        savedata.keys[3][SDID_JUMP]      = CONTROL_DEFAULT4_FIRE5; // chimera: keys, as player 1 has
+        savedata.keys[3][SDID_SPECIAL]   = CONTROL_DEFAULT4_FIRE6; // chimera: keys, as player 1 has
+        savedata.keys[3][SDID_START]     = CONTROL_DEFAULT4_START; // chimera: keys, as player 1 has
         savedata.keys[3][SDID_SCREENSHOT] = CONTROL_NONE;
         #ifdef SDL
             //savedata.keys[3][SDID_ESC]       = CONTROL_DEFAULT4_ESC;
diff --git a/engine/sdl/control.h b/engine/sdl/control.h
index 8593fb1b..8fc5313c 100644
--- a/engine/sdl/control.h
+++ b/engine/sdl/control.h
@@ -10,6 +10,7 @@
 #define	CONTROL_H
 
 // Generic control stuff (keyboard+joystick).
+// chimera: players 2 to 4 have keys by default, as player 1 does (no joysticks in the sandbox)
 #include <SDL_keycode.h>
 #include "joysticks.h"
 
@@ -28,45 +29,45 @@
 #define	CONTROL_DEFAULT1_SCREENSHOT	SDL_SCANCODE_F12
 #define	CONTROL_DEFAULT1_ESC        SDL_SCANCODE_ESCAPE
 
-#define	CONTROL_DEFAULT2_UP			((JOY_LIST_FIRST + 1) + JOY_MAX_INPUTS)
-#define	CONTROL_DEFAULT2_RIGHT		((JOY_LIST_FIRST + 2) + JOY_MAX_INPUTS)
-#define	CONTROL_DEFAULT2_DOWN		((JOY_LIST_FIRST + 3) + JOY_MAX_INPUTS)
-#define	CONTROL_DEFAULT2_LEFT		((JOY_LIST_FIRST + 4) + JOY_MAX_INPUTS)
-#define CONTROL_DEFAULT2_FIRE1		((JOY_LIST_FIRST + 5) + JOY_MAX_INPUTS)
-#define CONTROL_DEFAULT2_FIRE2		((JOY_LIST_FIRST + 6) + JOY_MAX_INPUTS)
-#define	CONTROL_DEFAULT2_FIRE3		((JOY_LIST_FIRST + 7) + JOY_MAX_INPUTS)
-#define	CONTROL_DEFAULT2_FIRE4		((JOY_LIST_FIRST + 8) + JOY_MAX_INPUTS)
-#define	CONTROL_DEFAULT2_FIRE5		((JOY_LIST_FIRST + 9) + JOY_MAX_INPUTS)
-#define	CONTROL_DEFAULT2_FIRE6		((JOY_LIST_FIRST + 10) + JOY_MAX_INPUTS)
-#define CONTROL_DEFAULT2_START		((JOY_LIST_FIRST + 11) + JOY_MAX_INPUTS)
+#define	CONTROL_DEFAULT2_UP			SDL_SCANCODE_I
+#define	CONTROL_DEFAULT2_RIGHT		SDL_SCANCODE_L
+#define	CONTROL_DEFAULT2_DOWN		SDL_SCANCODE_K
+#define	CONTROL_DEFAULT2_LEFT		SDL_SCANCODE_J
+#define CONTROL_DEFAULT2_FIRE1		SDL_SCANCODE_Q
+#define CONTROL_DEFAULT2_FIRE2		SDL_SCANCODE_W
+#define	CONTROL_DEFAULT2_FIRE3		SDL_SCANCODE_E
+#define	CONTROL_DEFAULT2_FIRE4		SDL_SCANCODE_R
+#define	CONTROL_DEFAULT2_FIRE5		SDL_SCANCODE_T
+#define	CONTROL_DEFAULT2_FIRE6		SDL_SCANCODE_Y
+#define CONTROL_DEFAULT2_START		SDL_SCANCODE_U
 #define CONTROL_DEFAULT2_SCREENSHOT ((JOY_LIST_FIRST + 12) + JOY_MAX_INPUTS)
 #define	CONTROL_DEFAULT2_ESC        ((JOY_LIST_FIRST + 15) + JOY_MAX_INPUTS)
 
-#define	CONTROL_DEFAULT3_UP			((JOY_LIST_FIRST + 1) + (JOY_MAX_INPUTS * 2))
-#define	CONTROL_DEFAULT3_RIGHT		((JOY_LIST_FIRST + 2) + (JOY_MAX_INPUTS * 2))
-#define	CONTROL_DEFAULT3_DOWN		((JOY_LIST_FIRST + 3) + (JOY_MAX_INPUTS * 2))
-#define	CONTROL_DEFAULT3_LEFT		((JOY_LIST_FIRST + 4) + (JOY_MAX_INPUTS * 2))
-#define CONTROL_DEFAULT3_FIRE1		((JOY_LIST_FIRST + 5) + (JOY_MAX_INPUTS * 2))
-#define CONTROL_DEFAULT3_FIRE2		((JOY_LIST_FIRST + 6) + (JOY_MAX_INPUTS * 2))
-#define	CONTROL_DEFAULT3_FIRE3		((JOY_LIST_FIRST + 7) + (JOY_MAX_INPUTS * 2))
-#define	CONTROL_DEFAULT3_FIRE4		((JOY_LIST_FIRST + 8) + (JOY_MAX_INPUTS * 2))
-#define	CONTROL_DEFAULT3_FIRE5		((JOY_LIST_FIRST + 9) + (JOY_MAX_INPUTS * 2))
-#define	CONTROL_DEFAULT3_FIRE6		((JOY_LIST_FIRST + 10) + (JOY_MAX_INPUTS * 2))
-#define CONTROL_DEFAULT3_START		((JOY_LIST_FIRST + 11) + (JOY_MAX_INPUTS * 2))
+#define	CONTROL_DEFAULT3_UP			SDL_SCANCODE_KP_8
+#define	CONTROL_DEFAULT3_RIGHT		SDL_SCANCODE_KP_6
+#define	CONTROL_DEFAULT3_DOWN		SDL_SCANCODE_KP_2
+#define	CONTROL_DEFAULT3_LEFT		SDL_SCANCODE_KP_4
+#define CONTROL_DEFAULT3_FIRE1		SDL_SCANCODE_KP_7
+#define CONTROL_DEFAULT3_FIRE2		SDL_SCANCODE_KP_9
+#define	CONTROL_DEFAULT3_FIRE3		SDL_SCANCODE_KP_1
+#define	CONTROL_DEFAULT3_FIRE4		SDL_SCANCODE_KP_3
+#define	CONTROL_DEFAULT3_FIRE5		SDL_SCANCODE_KP_0
+#define	CONTROL_DEFAULT3_FIRE6		SDL_SCANCODE_KP_PERIOD
+#define CONTROL_DEFAULT3_START		SDL_SCANCODE_KP_ENTER
 #define CONTROL_DEFAULT3_SCREENSHOT ((JOY_LIST_FIRST + 12) + (JOY_MAX_INPUTS * 2))
 #define	CONTROL_DEFAULT3_ESC        ((JOY_LIST_FIRST + 15) + (JOY_MAX_INPUTS * 2))
 
-#define	CONTROL_DEFAULT4_UP			((JOY_LIST_FIRST + 1) + (JOY_MAX_INPUTS * 3))
-#define	CONTROL_DEFAULT4_RIGHT		((JOY_LIST_FIRST + 2) + (JOY_MAX_INPUTS * 3))
-#define	CONTROL_DEFAULT4_DOWN		((JOY_LIST_FIRST + 3) + (JOY_MAX_INPUTS * 3))
-#define	CONTROL_DEFAULT4_LEFT		((JOY_LIST_FIRST + 4) + (JOY_MAX_INPUTS * 3))
-#define CONTROL_DEFAULT4_FIRE1		((JOY_LIST_FIRST + 5) + (JOY_MAX_INPUTS * 3))
-#define CONTROL_DEFAULT4_FIRE2		((JOY_LIST_FIRST + 6) + (JOY_MAX_INPUTS * 3))
-#define	CONTROL_DEFAULT4_FIRE3		((JOY_LIST_FIRST + 7) + (JOY_MAX_INPUTS * 3))
-#define	CONTROL_DEFAULT4_FIRE4		((JOY_LIST_FIRST + 8) + (JOY_MAX_INPUTS * 3))
-#define	CONTROL_DEFAULT4_FIRE5		((JOY_LIST_FIRST + 9) + (JOY_MAX_INPUTS * 3))
-#define	CONTROL_DEFAULT4_FIRE6		((JOY_LIST_FIRST + 10) + (JOY_MAX_INPUTS * 3))
-#define CONTROL_DEFAULT4_START		((JOY_LIST_FIRST + 11) + (JOY_MAX_INPUTS * 3))
+#define	CONTROL_DEFAULT4_UP			SDL_SCANCODE_1
+#define	CONTROL_DEFAULT4_RIGHT		SDL_SCANCODE_4
+#define	CONTROL_DEFAULT4_DOWN		SDL_SCANCODE_2
+#define	CONTROL_DEFAULT4_LEFT		SDL_SCANCODE_3
+#define CONTROL_DEFAULT4_FIRE1		SDL_SCANCODE_5
+#define CONTROL_DEFAULT4_FIRE2		SDL_SCANCODE_6
+#define	CONTROL_DEFAULT4_FIRE3		SDL_SCANCODE_7
+#define	CONTROL_DEFAULT4_FIRE4		SDL_SCANCODE_8
+#define	CONTROL_DEFAULT4_FIRE5		SDL_SCANCODE_9
+#define	CONTROL_DEFAULT4_FIRE6		SDL_SCANCODE_0
+#define CONTROL_DEFAULT4_START		SDL_SCANCODE_MINUS
 #define CONTROL_DEFAULT4_SCREENSHOT ((JOY_LIST_FIRST + 12) + (JOY_MAX_INPUTS * 3))
 #define	CONTROL_DEFAULT4_ESC        ((JOY_LIST_FIRST + 15) + (JOY_MAX_INPUTS * 3))
 
